larq/optimizers_v1.py

```diff
@@ -0,0 +1,180 @@
+        updates.extend(self.optimizer.get_updates(loss, base_lr_params))
+
+        return updates
+
+    def __getattr__(self, name):
+        return getattr(self.optimizer, name)
+
+    def get_config(self):
+        return {
+            "optimizer": {
+                "class_name": self.optimizer.__class__.__name__,
+                "config": self.optimizer.get_config(),
+            },
+            "multipliers": self.multipliers,
+        }
+
+    @classmethod
+    def from_config(cls, config, custom_objects=None):
+        optimizer = tf.keras.optimizers.deserialize(
+            config.pop("optimizer"), custom_objects=custom_objects
+        )
+        return cls(optimizer, config["multipliers"])
+
+
+@utils.register_keras_custom_object
+class Bop(tf.keras.optimizers.Optimizer):
+    def __init__(self, fp_optimizer, threshold=1e-5, gamma=1e-2, name="Bop", **kwargs):
+        super().__init__(**kwargs)
+
+        if not isinstance(fp_optimizer, tf.keras.optimizers.Optimizer):
+            raise TypeError(
+                f"Expected tf.keras.optimizers.Optimizer, received {type(fp_optimizer)}."
+            )
+
+        with K.name_scope(self.__class__.__name__):
+
+            self.fp_optimizer = fp_optimizer
+            self.threshold = threshold
+            self.gamma = gamma
+            self.iterations = K.variable(0, dtype="int64", name="iterations")
+
+    def get_updates(self, loss, params):
+        grads = self.get_gradients(loss, params)
+        self.updates = []
+
+        ms = [K.zeros(K.int_shape(p), dtype=K.dtype(p)) for p in params]
+
+        fp_params = []
+
+        for p, g, m in zip(params, grads, ms):
+            if "/kernel" in p.name and "quant_" in p.name:
+                print("Applying Bop to %s." % p.name)
+
+                m_t = (1 - self.gamma) * m + self.gamma * g
+
+                self.updates.append(state_ops.assign(m, m_t))
+                self.updates.append(
+                    state_ops.assign(
+                        p, lq.quantizers.sign(-p * tf.sign(p * m_t - self.threshold))
+                    )
+                )
+
+            else:
+                fp_params.append(p)
+
+        self.updates.append(state_ops.assign(self.iterations, self.iterations + 1))
+
+        return self.updates + self.fp_optimizer.get_updates(loss, fp_params)
+
+    @staticmethod
+    def is_binary(var):
+        return "/kernel" in var.name and "quant_" in var.name
+
+    def get_config(self):
+        fp_optimizer_config = self.fp_optimizer.get_config()
+        config = {
+            "threshold": self.threshold,
+            "gamma": self.gamma,
+            "fp_optimizer": {
+                "class_name": self.fp_optimizer.__class__.__name__,
+                "config": fp_optimizer_config,
+            },
+        }
+        return {**super().get_config(), **config}
+
+    @classmethod
+    def from_config(cls, config, custom_objects=None):
+        new_config = deepcopy(config)
```

Comment:
Do we really need a `deepcopy` here, or is `config = config.copy()` enough?

Reply:
need a deepcopy bc fp_optimizer is a nested dictionary.

Reply:
> need a deepcopy bc fp_optimizer is a nested dictionary.

True, but we only mutate the top level of the dictionary. Though we can leave it as `deepcopy` to be save :+1: